fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/security/KeyManager.java

```diff
@@ -23,46 +31,92 @@
   private final Wallet wallet;
+  private KeyPair keyPair;
 
   @Inject
   public KeyManager(AndroidKeysetManager keysetManager, Wallet wallet) {
     this.keysetManager = keysetManager;
     this.wallet = wallet;
 
     try {
+      cacheMainKey();
       Log.d(TAG, "Public Key = " + getMainPublicKey().getEncoded());
     } catch (IOException | GeneralSecurityException e) {
-      Log.e(TAG, "Failed to retrieve public key", e);
-      throw new IllegalStateException("Failed to retrieve public key", e);
+      Log.e(TAG, "Failed to retrieve device's key", e);
+      throw new IllegalStateException("Failed to retrieve device's key", e);
     }
   }
 
-  public PublicKey getMainPublicKey() throws IOException, GeneralSecurityException {
-    return getPublicKey(keysetManager.getKeysetHandle());
+  /**
+   * This will cache the device KeyPair by extracting it from Tink.
+   *
+   * <p>Use this only if you know what you are doing
+   *
+   * @throws IOException when the key cannot be retrieved due to IO errors
+   * @throws GeneralSecurityException when the retrieved key is not valid
+   */
+  private void cacheMainKey() throws GeneralSecurityException, IOException {
+    keyPair = getKeyPair(keysetManager.getKeysetHandle());
   }
 
-  public KeyPair getMainKeyPair() throws IOException, GeneralSecurityException {
-    PrivateKey privateKey = new ProtectedPrivateKey(keysetManager.getKeysetHandle());
-    PublicKey publicKey = getMainPublicKey();
+  /** @return the device public key */
+  public PublicKey getMainPublicKey() {
+    return keyPair.getPublicKey();
+  }
 
-    return new KeyPair(privateKey, publicKey);
+  /** @return the deice keypair */
+  public KeyPair getMainKeyPair() {
+    return keyPair;
+  }
+
+  /**
+   * Try to retrieve the user's PoPToken for the given Lao.
+   *
+   * @param lao we want to retrieve the PoP Token from
+   * @return the PoP Token if it was retrieved, empty if no RollCall exist in the LAO
+   * @throws KeyGenerationException if an error occurs during key generation
+   * @throws UninitializedWalletException if the wallet is not initialized with a seed
+   * @throws InvalidPoPTokenException if the token is not a valid attendee
+   * @throws NoRollCallException if the LAO has no RollCall
+   */
+  public PoPToken getValidPoPToken(Lao lao) throws KeyException {
+    // Find the latest closed RollCall and use the wallet to retrieve the key
+    RollCall rollCall =
+        lao.getRollCalls().values().stream().max(Comparator.comparing(RollCall::getEnd))
+            .orElseThrow(() -> new NoRollCallException(lao));
+
+    return getValidPoPToken(lao, rollCall);
   }
 
-  public PoPToken getPoPToken(String laoID, String rollCallID) throws KeyException {
-    return wallet.generatePoPToken(laoID, rollCallID);
+  /**
+   * Try to retrieve the user's PoPToken for the given Lao and RollCall. It will fail is the user did not attend
+   * the roll call of if the token cannot be generated
```

Comment:
```suggestion
   * Try to retrieve the user's PoPToken for the given Lao and RollCall. It will fail if the user did not attend
   * the roll call or if the token cannot be generated
```